src/commands/find-buf.ts

```diff
@@ -0,0 +1,143 @@
+
+      if (configPath) {
+        if (!path.isAbsolute(configPath)) {
+          configPath = path.join(
+            vscode.workspace.rootPath || process.cwd(), //todo: fix me in the future when we want to support multi-root workspaces
+            configPath
+          );
+        }
+
+        if (configVersion) {
+          log.warn(
+            "Both 'buf.commandLine.path' and 'buf.commandLine.version' are set. Using 'buf.commandLine.path'."
+          );
+        }
+
+        try {
+          log.info(`Buf path set to '${configPath}'.`);
+          bufCtx.buf = await BufVersion.fromPath(configPath);
+
+          if (bufCtx.buf) {
+            log.info(
+              `Using '${bufCtx.buf.path}', version: ${bufCtx.buf.version}.`
+            );
+          }
+        } catch (e) {
+          log.error(
+            `Error loading buf from path '${configPath}': ${unwrapError(e)}`
+          );
+          bufCtx.buf = undefined;
+        }
+
+        return;
+      }
+
+      if (configVersion) {
+        try {
+          log.info(
+            `Buf version set to '${configVersion}'. Looking for it in the extension storage...`
+          );
+          bufCtx.buf = await findBufInStorage(ctx, configVersion);
+
+          if (bufCtx.buf) {
+            log.info(
+              `Using '${bufCtx.buf.path}', version: ${bufCtx.buf.version}.`
+            );
+          }
+        } catch (e) {
+          log.error(`Error finding an installed buf: ${unwrapError(e)}`);
+          bufCtx.buf = undefined;
+        }
+
+        return;
+      }
+
+      // If we didn't get a valid buf binary from config...
+      log.info("Looking for buf on the path...");
+
+      // If we didn't find it in storage...
+      try {
+        bufCtx.buf = await findBufInPath();
+
+        if (bufCtx.buf) {
+          log.info(
+            `Using '${bufCtx.buf.path}', version: ${bufCtx.buf.version}.`
+          );
+          return;
+        }
+      } catch (e) {
+        log.error(`Error finding an installed buf: ${unwrapError(e)}`);
+      }
+
+      // If we made it to this point, we found nothing, throw an error.
+      log.error(
+        "Buf is not installed or no valid binary could be found. Please install Buf."
+      );
+    };
+  }
+);
+
+const findBufInPath = async (): Promise<BufVersion | undefined> => {
+  const bufPath = await which(bufFilename, { nothrow: true });
+
+  if (bufPath) {
+    return BufVersion.fromPath(bufPath);
+  }
+
+  return undefined;
+};
+
+const findBufInStorage = async (
+  ctx: vscode.ExtensionContext,
+  version: string
+): Promise<BufVersion | undefined> => {
+  const files = await fs.promises.readdir(ctx.globalStorageUri.fsPath);
+
+  // If the user has not set a specific version, we do nothing.
+  if (!version) {
+    return undefined;
+  }
+
+  let found: string | undefined;
+
+  if (version === "latest") {
+    found = files
+      .filter((f) => f.startsWith("v"))
+      .sort()
+      .at(-1);
```

Comment:
This behaviour doesn't seem correct -- if the user has anything installed in their global storage, it would be used as `latest`, even if it is not actually the latest version of the CLI... so in the case where the user has formerly installed `v1.40.0` for example, and then changes the value to `latest`, no update would actually happen.

Reply:
Changed the handling per @paul-sachs's approach slightly modified.